### ISSUE TITLE
Add Developer menu and functions

### ISSUE THREAD

**Issue by Darknior** [gazpan/WebtroPie#40]
Ideas to help Skin developers with Webtropie.

> yes, maybe like the media preview when you mouse over a text, image, rating etc (holding down a special key or after setting a mode) then maybe the border is dotted (like uploading) and there's info of dimensions, position, z-index (like preview) and either something to click to copy xml to clipboard / show bubble or right click ?

> ... a little worry is though, that it's implying that WebtroPie is 100% accurate at displaying themes but it isn't there yet so I'd have to warn people that it's just for help and not definitive.

**Comment by gazpan** [comment]
I've added a start to this, at the moment it just shows the information, it doesn't work for game list or carousel yet and doesn't work well with small text.
The area the pops up is scrollable but think I need to do it a different way for small objects, like a second bubble we talked about.

**Comment by Darknior** [comment]
I try it and it's excellent :)
It's a good beginning ... it will really help us to work on new themes :)
Yes for the little elements, it can be cool to have a little popup where you write informations.
Maybe you can also hide the scrollbar without removing the scrolling capability ...

It is my proof of concept i work on today ... hard to find good ideas ...
![rvgm skin](https://user-images.githubusercontent.com/3703317/35466705-951435ce-0306-11e8-8b45-62d3ba875c36.jpg)

**Comment by gazpan** [comment]
wow, that looks excellent!, I like dark steel for large artwork this will be great too.

yes, I think a popup (draggable) could show all attributes, the name and input boxes for the values. For position and size these would change as the element is dragged and resized and also could be manually entered but for others like alignment and font size these could be input and have the changes applied to the element, then perhaps a button at the bottom to export xml to clipboard.

... it might be worth having a "snap to" adjustable grid to make it easier to align elements with a way of adjusting the grid size, perhaps 1% to 10% and off.

Maybe, in the future I will think again about saving, it might not be too bad now I've thought about it a bit more.

**Comment by Darknior** [comment]
Yes popup with all informations and maybe input to change them is really a good idea and making possible to copy XML code.
Yes i also have think to this GRID option to make easy to align element ... it can help a lot but we can also adjust with input by copying from one element to an other ;)
My projet is only an PNG for the moment, it can be interesting if i can use Webtropie to make it real and try what is working fine or not and what we must improve or not...

![rvgm skin 3do](https://user-images.githubusercontent.com/3703317/35474804-a31b8ab8-0393-11e8-8ab8-2c45a0cfa1ab.jpg)
![rvgm skin psx](https://user-images.githubusercontent.com/3703317/35474805-a4c7a932-0393-11e8-9eab-2add5f5b9330.jpg)
![rvgm skin snes](https://user-images.githubusercontent.com/3703317/35474806-a696e048-0393-11e8-81f0-1549029fb131.jpg)

**Comment by Darknior** [comment]
But i will improve some detail with working on FONT to add some cool icons i want to add to my Skin.
For exemple on Player Number i want to show 1-2 and not "Vs" but an VERSUS ICON :D 
And an other font for the release date to make it more beautiful ...
It's hard to find the good ideas :p

**Comment by gazpan** [comment]
I really like the look of your theme, yes, if you can use WebtroPie while making it that would be great for ideas and making it do what it needs to do.

I've done a little more now, elements can be resized and moved and there is a property popup... still lots to do on it but I thought I would commit at this point even though it's still very much WIP.

I think I need to maybe keep the property popup open after a click and have it draggable, I still need to get game list and carousel, to show up when moused over

... also need to add button for copy xml, make inputs for  the property palette and give it a proper header containing the name (just more styling).

... and down the road there will need to be a way to add new elements

**Comment by gazpan** [comment]
I was thinking about the grid too, I thought that maybe the user could select the resolution that the theme is targeted at because I think that grid alignment should at least try to help position on whole pixel boundaries ... or am I overthinking it ?, I would imagine that if it was possible to display an image exactly as it was edited then that would be preferred than it being stretched or aliased because it co-ordinates are half way between pixels or it's size is wrong.

So, I thought or offering typical pi resolutions, ... or would most people just use 1024 x 768 ?, If I did give a choice then I could try to set the window size to those exact dimensions (although I think some/most browsers might not allow it) and then, while dragging or resizing I could ensure that positions and sizes are proportions that work out to whole numbers (exact pixels) when it's multiplied by the resolution.

Do themers target a resolution ?

So, if I did have a grid size of 1024 x 768 or some other resolution then I would offer snap sizes of say 5, 10, 15, 20 pixels

Or, I could just not try and offer snap sizes in percentage units 1%, 2%, 4% ... ?

**Comment by Darknior** [comment]
WOAW if you can do all of that it will be perfect to try to make a Theme :)
Moving and resize element will help a lot.
Add an element to, but add it after, for start maybe we can add it on the XML our self and refresh.
Yes let the popup always open, if we want to close it we click on [X]
Maybe add a GRID overlay can help to align elements. With a PNG image ? But maybe we can't do it because if there is an element on top of the skin we can't interact with elements.

... i understand it will be long ... but incredible too :p ;)

**Comment by gazpan** [comment]
... yes, I might need a popup just for the grid, so that the z-index could be adjusted as well as the grid/snap size and to turn it on/off.

i've committed the changes that allow move/resize but haven't done grid yes. I think I'll have the image generated with a canvas dynamically as the grid live spacing could change and the overall size

**Comment by Darknior** [comment]
Yes you are right for the grid. it is important to aligne elements.
And if you want to give a GRID with the Skin size it is possible, the internet browser will add scrollbar to let us move.

1024 x 768 why not ... on TV we most use 720 or 1080p ! I try 4K but all is slow in 4K with PI3 lol
1080 x 720 or 1920 x 1080 ... my skin is in 1080p and works fine too in 720p. Because all is in %

I will try your commit now :)

**Comment by gazpan** [comment]
enjoy! ... I'll have a look tomorrow, bonne nuit

**Comment by Darknior** [comment]
And don't forgot my translation update please :p

**Comment by Darknior** [comment]
It is really excellent lol
To see elements info in popup and moving them, resize them, WOAW !!!!
I know it is a beta but i found a funny bug lol when i resize the video it stay with the new saved size, but on my system logo for exemple, i see it with the good new size but the over always show the old original size lol
When you will have add the gaming list, the zindex support etc ... you will have create the best ever skin editor<3

**Comment by gazpan** [comment]
thanks, I've merged the translation now
I found a couple of small bugs with resizing that I've fixed now, hope that has solved the problem.
Also it's a bit confusing when accidentally picking up an image behind another image ... like the logo2 behind the video ?

**Comment by Darknior** [comment]
> Also it's a bit confusing when accidentally picking up an image behind another image ... like the logo2 behind the video ?

No it's not, it's good for me. How do you want to do ?

I always have the bug : 

![2](https://user-images.githubusercontent.com/3703317/35503118-78e727f0-04df-11e8-84c5-533da5d5b701.jpg)
![1](https://user-images.githubusercontent.com/3703317/35503123-7baeeea0-04df-11e8-9fbb-0a514cb87c98.jpg)

Like on my screenshot if you can also show PIXEL size it will help us when we design it. On Photoshop i work with pixel, and i know the pixel size of my elements.

**Comment by Darknior** [comment]
One of the best ever online editer to make web site **moqups**.
https://app.moqups.com/edit/page/ad64222d5

You can't do the same, so many years of work ... 
ES has no many options too.
The flying popup is a good idea a think : 

![popup](https://user-images.githubusercontent.com/3703317/35508199-bf0cafa6-04ef-11e8-98dc-9f26e71518bb.jpg)

And you must add some options for TEXT based on what ES let us modify : 

![popup2](https://user-images.githubusercontent.com/3703317/35508223-d54b57ea-04ef-11e8-875f-748cd97be98a.jpg)

**Comment by Darknior** [comment]
I have try to work with my model and it is hard to use ... the editor help but there are some bugs...

![rvgm skin](https://user-images.githubusercontent.com/3703317/35540532-9db6b42a-0556-11e8-8b68-2f28a767a49c.jpg)

EDITOR BUGS : 

- The resize bug come from the element with a maxSize and not size, i can resize them but after the size is not saved ... i don't have this problem with an element that only have a size. To work on these elements i give them a size, work to find the real size, and after i switch size to maxSize in the XML.
- When i want to resize an element with maxSize it is strange it only resize with a ratio effect ... really hard to use :(
- If an element is put at the same place of the information bubble, i can't see it, you must increase the z-index.

WEBTROPIE BUG : 

- When i refresh the explorer, it don't stay on my selected VIEW :(
Maybe because i select a new VIEW not BASIC, DETAILED, VIDEO, i want to create some news ... like GRID in the future ... it is not important if we can't select them by default in options because you don't know them, but can you save the selected one to prevent explorer refresh ? Thanks

It help me a lot to do it for the moment :)
Seeing pixel size in the future will help a lot.

![rvgm skin 2](https://user-images.githubusercontent.com/3703317/35544044-a959de80-0568-11e8-86f7-601c3d2d0394.jpg)

**Comment by Darknior** [comment]
- Can we use VARIABLE with your skin engine ?
I have try to use them, maybe i do a mistake, but Webtropie can't read my XML with them.
It is a really cool feature nobody use for System path and colors ...

1. System Variables
2. Theme Defined Variables

https://github.com/RetroPie/EmulationStation/blob/master/THEMES.md

**Comment by Darknior** [comment]
- Think to **hide scrollbar** on gamelist.
`class="gamelist_outer" -> overflow: hidden;`

- I have an error on the **gamelist**, when i align it to **RIGHT** i have a space on RIGHT near the scrollbar.

> 		<alignment>right</alignment>
> 		<horizontalMargin>0.0</horizontalMargin>

`width: 21.64vw; -> width: 22vw; and same for max-width`
I don't know why there is the calcul error on RIGHT and not left, maybe for the Scrollbar.

- When i have try to **center images with POS 0.5 0.5** it works fine, i think you have a bug when we want to do same with **VIDEO**.
-

**Comment by gazpan** [comment]
I think that the editor bugs from a couple of posts ago should be ok now, I did a commit but it's in the middle of changes really. 

The property palette isn't two-way binding, changes in the palette don't update the element.

There's lots to do on this, I will refer back to the other issues when I've done a bit more.

Also, doing the property palette has highlighted a few attributes that I haven't coded yet, so even more to do.

The positions and sizes doesn't show pixels yet, but for that it needs to know what resolution to multiply by but I have a plan of how this will all work.

I will add a grid popup and offer the three main resolutions that you mentioned above, each on a different line, each a different colour : red, green, blue.
Each can have a different grid size (enterable or radio/select) and create a grid for it's resolution
Each also has a checkbox to turn it on or off, which also shows dimensions in the property palette colour coded to that resolution, so if you have two resolutions ticked, you would see two lots of pixel positions and sizes.

When each grid is turned on they have an alpha, when grid lines coincide it is good, it means that we are on a whole pixel boundary for multiple resolutions so edges will look sharp for both resolutions.

As well as the three resolutions I will also offer a fractions grid, which just shows halfway lines (strong), quarter lines (less alpha), eighths (even less alpha).

All resolutions are divisible be 4, 8 ... I haven't done the maths yet, so these are good lines to align to.

Snap will snap while dragging to the nearest line if which ever of these is switched on - slightly more priority to the fractions lines.

I can imagine this all sounds over complicated but I think when it's on screen it will make sense.

When all the basics are there i'll polish things up so there are better controls on the property palette for example - change checkboxes to styled images, justification to icons etc.

.... it's still really rough for now but hope the fixes I made help

... there's also issues when navigating to different pages with popups switched on

**Comment by gazpan** [comment]
... I forgot to add one file , it's there now

**Comment by Darknior** [comment]
Ok lol it's better with it ... Ouf :D 
- Now i must select an element before working with it, cool ... ok. but if it is little i can't close it !
Maybe you can also add the [X] on the informations popup to close it :)

I will try it more ... but for the moment i love it ... and yes i know you have many work with all the elements attribute you don't use for the moment, and the VARIABLE i think you don't use too.
I work with your editor and the XML at same time and listen how to make a good Theme with good code :)
I want to make it with almost 10 different VIEW ... it's why i report you this bug.

Please can you just update the bug i report before : 
- When i refresh the explorer, it don't stay on my selected VIEW :(
- And remove the gamelist scrollbar ?

**Comment by Darknior** [comment]
On datetime you forgot some attribute : 

color - type: COLOR.
fontPath - type: PATH.
fontSize - type: FLOAT.
forceUppercase - type: BOOLEAN. Draw text in uppercase.

And i don't want to give you more work :(
So bad i can't help you more ...
But after you can add the option informations like on the online doc. Maybe with a title on input over ?
I can translate all of it in french if you make it possible.

**Comment by gazpan** [comment]
... yes, thanks I got about halfway through the online documents, down to text list I think.
There is some support for variables, the themes that use variables all work but I don't think it's 100% for every type of variable on all elements. ... I have to go out now so I may not be able to update tonight
:(

**Comment by Darknior** [comment]
Not a problem, maybe tomorow ;)
Thanks, have a good evening :)

**Comment by gazpan** [comment]
.. oh, there is an X on the image overlay ... good idea to put it on properties too

**Comment by gazpan** [comment]
thanks :)

**Comment by Darknior** [comment]
I work a lot to understand fine all the feature possible for teaming on ES but some are strange lol

Some time i have a CRASH like this one

```
angular.min.js:123 TypeError: Cannot read property 'image' of undefined
    at theme.service.js:499
    at Object.p [as forEach] (angular.min.js:8)
    at theme.service.js:495
    at Object.p [as forEach] (angular.min.js:8)
    at theme.service.js:493
    at Object.p [as forEach] (angular.min.js:8)
    at findStaticImages (theme.service.js:491)
    at decodeTheme (theme.service.js:480)
    at onSuccess (theme.service.js:642)
    at angular.min.js:136 "Possibly unhandled rejection: {}"
```
Maybe it's possible to have a better LOG error on console with the name and the line of the XML that crash your code ?

And i have some difficult with the RATING ... i replace the original STARS by some other but it not working evey time, some time it only take the color of my image like a texture and show the original star ... really strange :(

And i think i must debug my skin with ES on PI because it works fine on WEBTROPIE and not on PI ... strange too.

**Comment by Darknior** [comment]
I found an other error with the RATING element.

![error2](https://user-images.githubusercontent.com/3703317/35635142-c6d81c6a-06ad-11e8-931a-66c8ab0032ba.jpg)

Like you can see on WEBTOPIE you always add the YELLOW color to the element. If you can remove it like on ES (bottom image)

![error1](https://user-images.githubusercontent.com/3703317/35635147-c9ed9254-06ad-11e8-81a8-25b068d95c28.jpg)

And i discover like for the HELP BAR, the ELEMENT LABEL are show on ES if we don't use them on our XML. We must declare them and use position "2 2" to hide them over the screen.
A problem we don't have on WEBTROPIE ;)

**Comment by tomaz82** [comment]
No no no no no, thumbnail isn't a new attribute, nothing was added, the fix was that in video list md_image used the same texture as thumbnail, when it should have used the image texture. Nothing new was added.

**Comment by Darknior** [comment]
> No no no no no, thumbnail isn't a new attribute, nothing was added, the fix was that in video list md_image used the same texture as thumbnail, when it should have used the image texture. Nothing new was added.

Yes sorry for my bad English, i understand this fine.
But i think @gazpan has not implement thumbnail for the moment because they are not explained on the theme doc ...
https://github.com/RetroPie/EmulationStation/blob/master/THEMES.md

I think it would be good to add it after this line : **showSnapshotNoVideo - type: BOOLEAN**

> If there is no video to show, ES will show the `<thumbnail>` image, and if it's empty or not exist, using the `<image>` from your gamelist.xml

**Comment by Darknior** [comment]
Hi @gazpan ... no news about you from 15 days now ....
I was sick myself last week and I still am .... but i try when i can to work on my skin ... i will show it to you in a few time ;)
Are you ok ? All is fine in your life ?
Have a good week-end ;) Bye

**Comment by gazpan** [comment]
Hi, sorry for no updates or commits,I've been busy with other RL stuff and progressing the theme editor from time to time, there are lots of code changes to tidy up and I would like to do more and make it all stable before I push it next.

About the view problem that you had after a refresh, you can set the view in main menu ui options (to something other than automatic) and it will stay on that view, the option on the helpbar only temporarily changes the view, but, perhaps this should be remembered?

I get that it would be useful, to remember the view for that system for that theme but what if you want undo this setting?, ... maybe a "[ ] Save" a last option (checkbox) appears if you change the view that you can then click or unclick...

**Comment by Darknior** [comment]
Really happy to read you @gazpan :)
I don't search for more updates or commits ... only to know you are fine. I just found it strange that you disappear without a word 15 days lol ... I know that real life can sometimes make bad jokes :(

- 4 post ago i list all the bugs i can found when i try to create my new theme, to help you to fix them.
- And i create some new ISSUES for the Webtropie bugs i found.

My skin is near finished, i will show it to you, i use some trick to make it with your editor, but make it without Webtropie were really be a pain :(
Now i think i understand 100% how working Emulationstation Skinning and it will help us to debug Webtropie :)

> About the view problem that you had after a refresh, you can set the view in main menu ui options (to something other than automatic) and it will stay on that view, the option on the helpbar only temporarily changes the view, but, perhaps this should be remembered?

You can forgot this BUG, it's not one. It's why i remove it from my BUG LIST 4 post before.
I understand now ES only can show 4 VIEW. And we don't have the choice, each view has a name hard coded :( and habilitiez hard coded too :( A really bad system ...
I will never use Basic and Detailled view ... the only one i like is Video view and i can't create other video view with some difference .... ans now the GRID view come ... but i'm not really fan too :(
So bad if we can't find a cool guy, one day, that know C++ to upgrade the system view of ES ...
To make it possible to have many view, where we can use the attributes we choose, all with video if we want, but with rom browser left or right, with more or less detail, bigger or smaller video etc ...

Your only real bug is that in the menu you don't read the theme XML and always list : **Automatic / Basic / Detailled / Video**
If my theme also use GRID you forgot it. And i can't make it default.

**Comment by gazpan** [comment]
Glad that you were able to make your theme without me progressing much with the editor.
I've pushed some changes now, sorry it's not more but I haven't given up, so, in time we will get there.

The menu always offers Automatic, basic, detailed and video as it's like a target view to try for any given theme. The view selector on the helpbar switches between the actual views available for the current theme.

The user might want to select 'video' as a default even if the current theme doesn't have a video.

It supposed to try to set the view to the selected view unless it doesn't exist for the current theme/system and if not it's sets it to video if there are videos, detailed if there are images otherwise basic.

Although the helpbar shows 'grid' view and any other view contained in the theme the grid logic hasn't been done yet so it won't work properly at the moment, this is why I haven't added it as a default option yet.

I'm happy to change any of this but not sure what the alternative logic should be ?

**Comment by Darknior** [comment]
> I'm happy to change any of this but not sure what the alternative logic should be ?

No, change nothing, it's fine like this. Only add GRID in the future when it will works fine. That's all :)

Thanks for your new update :)
- Possible to unselect element by the detail window :)
- Possible to modify values on the detail window and see the change (not all on gamelist for exemple)
- Add of near all the proprieties of elements
- Add combo-box with type and name of element
And maybe some other i don't see ....

Great work .... thanks 👍

**Comment by gazpan** [comment]
yes, I will work on getting all of the attributes editable for all types of element, it's easy to say but a little slow and boring to do.. :( ... but it's worth it so I will be glad when it is done.

A few of the things I came across I hadn't implemented, like rotation and rotation origin, it was cool to play around with those figures in the property palette on text or images, I'm not sure if it does the same as ES yet, I need to test but it was fun to rotate text and images. I image in would be difficult to use things like that without having an editor because it would be very slow to try over and over restarting ES, ... maybe there will be more themes that use more of the features available in future if this get used,

... good ideas about conbo boxes and moving elements with the keyboard.

I will do more this weekend, there still so many things that I want to do, it might be a bit random whatever I end up doing. Have a nice weekend.

**Comment by Darknior** [comment]
I know it is really a HARD work you do :(

> A few of the things I came across I hadn't implemented, like rotation and rotation origin, it was cool to play around with those figures in the property palette on text or images

Yes i love this feature, it is really fun to use :)
And now easier with your Webtropie editor :D 

I don't know if it will be many more rotation used on Theme because if i understand fine there is always the MEMORY problem :( We can't use so many elements ....
On my actual theme i use many little images to separate elements to make the theme customizable and let us choose language easily ... And some time all the images are disappear :(
I will try to make some test with less images to see if it's fix the problem.

> ... good ideas about conbo boxes and moving elements with the keyboard.

- Combo box is your idea :)
I love it but there is a CSS problem, i can't read them fine when open.
Write in white on white background lol

![photo](https://user-images.githubusercontent.com/3703317/36339962-d45f3a96-13d0-11e8-93cb-867b8af76d0f.jpg)

- Moving element with keyboard will help to gain precision :)

Thanks a lot again for all your hard work :D
